### DISCUSSION TITLE
Display array of lines and arcs / Viewport touch support

### DISCUSSION THREAD

**Discussion by borioan** [Macad3D/Macad3D#3] [Ideas]
I have two basic question about using Macad:
1. I need to draw/display inside the main Macad3D window an array/(or maybe a) list of 3D lines and arcs - it could be few hundreds of Kb or even up to 50 Mb of these lines and arcs. Usually, to draw these lines in an OpenGl window, I used the standard OpenGL draw syntax: 
GL.Begin(GL.LINES);
GL.Color(lineColor);
int i;
for (i = 0; i < linePoints.Count; i++)
{
     GL.Vertex(linePoints[i]);
}
GL.End();
It is possible to use the direct method above in Macad or there are specific data structure in OpenCascade to help me with this problem?  Browsing your code I have seen that there is an AIS_InteractiveObject class. I'm not sure if this can help me to display the 3D lines and arcs I need it. Let me know what is the recommended method for this first issue, the best is to have a simple example. Please note that the lines are just for visualize/display it there are not selectable, but of course I must be able to erase them if I need to display other different array of lines and some memory management must exist to clean up the viewport correctly. Also I must have the option to use different colors for these lines.
2. it is possible to add support for visualize the 3d parts using **touch event**?  I have studied Macad code source and I discovered it is heavily dependent on mouse events, but I need the touch only for visualize the 3d models - pan zoom and  rotate 3D. 
Thank you very much!!

**Comment by dipts** [comment]
1. There are classes with which tools can show auxiliary lines, but it doesn't scale over that amount. The best way would actually be via an AIS_InteractiveObject derivation, which allows the use of OCCT drawing objects or direct drawing with OpenGL. Examples of this can be found in the _Macad.Managed_ project or in the [OCCT documentation](https://dev.opencascade.org/doc/overview/html/occt_user_guides__visualization.html#occt_visu_4_2).
2. Sure it would be possible. The mouse events are handled directly in the UserControl (_ViewportHwndView_, _ViewportViewModel_); the touch events should then also be handled at this point. Both create the same or very similar movement patterns, maybe it is enough to trigger the corresponding mouse move functions on the _ViewportController_. It should be easy to try.

**Comment by dipts** [comment]
For 1. there is another possibility: You can create edges from the lines and arcs and then throw them into a body using a "solid" shape. For coloring, there is currently only the option of sorting it into individual layers.
```csharp
Geom2d_Curve curve1 = // Create curve for line or arc
TopoDS_Shape edge1 = new BRepBuilderAPI_MakeEdge2d(curve1).Edge();
Geom2d_Curve curve2 = // Create curve for line or arc
TopoDS_Shape edge2 = new BRepBuilderAPI_MakeEdge2d(curve2).Edge();

// Create BRep
var builder = new TopoDS_Builder();
var shape = new TopoDS_Compound();
builder.MakeCompound(shape);
builder.Add(shape, edge1);
builder.Add(shape, edge2);

// Create body
var solid = Solid.Create(shape);
var body = Body.Create(solid);
Document.AddChild(body);

// Create new layer
var layer = new Layer()
{
    Name = "MyLayer",
    Color = new Macad.Common.Color("#ff0000"),
    LineStyle = LineStyle.Dash,
    LineThickness = LineThickness.Thin
};
Layers.Add(layer);

// Sort body into layer
body.Layer = layer;
```

**Comment by borioan** [comment]
Thank you very much!!!!!!!!!!! I will try the second option, first time and come back if I have no success!!!
It seems to be the most easy way!
The only problem is that the lines and arcs are 3D, I'm not sure how can I fit them in a Geom2d_Curve type..
In fact for lines all I have is 3D points - an array - identical with the OpenGL example from the starting of the discussion..

**Reply by dipts** [reply]
You can create Geom_Curve, it's the same in 3D. [Check the OCCT guide for this.](https://dev.opencascade.org/doc/overview/html/occt_user_guides__modeling_data.html#occt_modat_3)

**Reply by borioan** [reply]
Thank you!
I hope to be efficient in memory terms in the case the points files are bigger  (Mb..)

**Comment by borioan** [comment]
Hi!
It is possible to save the body using the "solid" shape as you indicated? I have successfully draw the lines (first try) but unfortunately it seems that it can not be saved... :(  
For the first possibility (as AIS_InteractiveObject derivation) I still have no clue how to start, some help will be much appreciated!
Speaking about the speed calculation, using this method with fake body, for 5 Mb of lines it take approximately 1 minute (in Debug mode!) to display it on the main window. It works, but less time is better :-) The computer has the following configuration: i9 9900/32 Gb RAM, 512 Gb Optane Intel SSD

**Reply by dipts** [reply]
You need to add the body to the model: `Document.AddChild(body);` in script, `CoreContext.Current.Document.AddChild(body);` if you're in app code. Then it is part of the model and will be saved.

Never measure execution times in debug build, the deviations from the release build are enormous. Note that the entire OCCT library also runs unoptimized in the debug build. If the time is still too long, you should profile where the time remains and optimize there. It may help to relocate the creation to managed code, where OCCT can be used more efficiently and the transitions between the managed and native layers are reduced.

**Comment by borioan** [comment]
Excellent , excellent, excellent!! It works perfectly and the file size is just the right one! 
Regarding the release build, I encountered an issue with running the release version. It is explained [here](https://github.com/Macad3D/Macad3D/discussions/5).
Before taking care of the touch event, I just want to ask if you can direct me how to create programmatically a new layer, setup its properties (color, transparency, line width, etc.) and put the hole bunch of the lines on it. Also I need a new layer for importing a STEP file.
I love Macad3D !! :-)

**Reply by dipts** [reply]
Ok, I'll take a look into that problem.

To create a layer:
```cs
var layer = new Layer()
{
    Name = "MyLayer",
    Color = new Macad.Common.Color("#ff0000"),
    LineStyle = LineStyle.Dash,
    LineThickness = LineThickness.Thin
};
Layers.Add(layer);

body.Layer = layer;
```
I add it to the script in the first answer, to get a complete solution there.

For the touch events, it would be best to open a new discussion for that.

**Comment by borioan** [comment]
Perfect!! For touch will be a different / new discussion!
One last important question about layers before closing this discussion: some of my lines are displayed in a color, the other are colored differently. It is possible in some way to "subclass" a layer, in other words, in one "main" layer to insert to different "sub"layers (or more) which represent these 2 group of lines? Or maybe you can suggest some kind of visually grouping together these two sets of lines.  
Most of the time both types of lines are consider to be part of a singe entity or body in Macad approach. The only difference between these two groups of lines are the color. When I want to hide both of them, I want the user to not feel the differences and to push two times for hide both of them. Any idea/workaround will be great! Also I notice that there is a different speed of hiding them: using the layer eye check button is much faster (instantly!) than the eye button in the Document explorer, so for big data lines the layer approach is the best!
Thank you very much!

**Reply by dipts** [reply]
There is an unfinished feature for overwriting visual properties per body, but there is still a lot of work involved, including creating quite a few tests.

One possibility would be to switch the layers dependent on each other. `Layer` implements the `PropertyChanged` event, which can be subscribed to and if the `IsVisible` property is changed, the visibility of the other layer can be adjusted.

> Also I notice that there is a different speed of hiding them: using the layer eye check button is much faster (instantly!) than the eye button in the Document explorer

This is quite interesting, I'll have a look at it sometime.
